core/rawdb/freezer.go

```diff
@@ -336,62 +361,109 @@ func (f *Freezer) Sync() error {
-		tail = table.itemHidden.Load()
-		name = kind
-		break
-	}
-	// Now check every table against those boundaries.
-	for kind, table := range f.tables {
-		if head != table.items.Load() {
-			return fmt.Errorf("freezer tables %s and %s have differing head: %d != %d", kind, name, table.items.Load(), head)
+
+	check := func(tables map[string]*freezerTable) (uint64, uint64, error) {
+		var (
+			head uint64
+			tail uint64
+			name string
+		)
+		// Hack to get boundary of any table
+		for kind, table := range tables {
+			head = table.items.Load()
+			tail = table.itemHidden.Load()
+			name = kind
+			break
 		}
-		if tail != table.itemHidden.Load() {
-			return fmt.Errorf("freezer tables %s and %s have differing tail: %d != %d", kind, name, table.itemHidden.Load(), tail)
+		// Now check every table against those boundaries.
+		for kind, table := range tables {
+			if head != table.items.Load() {
+				return 0, 0, fmt.Errorf("freezer tables %s and %s have differing head: %d != %d", kind, name, table.items.Load(), head)
+			}
+			if tail != table.itemHidden.Load() {
+				return 0, 0, fmt.Errorf("freezer tables %s and %s have differing tail: %d != %d", kind, name, table.itemHidden.Load(), tail)
+			}
 		}
+		return head, tail, nil
+	}
+
+	headTables, otherTables := splitTables(f.tables)
+	// verify that all the non-header tables have the same size
+	head, tail, err := check(otherTables)
+	if err != nil {
+		return err
 	}
+	// verify that the header tables have the same size
+	if len(headTables) != 0 {
+		head, tail, err = check(headTables)
+		if err != nil {
+			return err
+		}
+	}
+
 	f.frozen.Store(head)
 	f.tail.Store(tail)
 	return nil
 }
 
 // repair truncates all data tables to the same length.
 func (f *Freezer) repair() error {
-	var (
-		head = uint64(math.MaxUint64)
-		tail = uint64(0)
-	)
-	for _, table := range f.tables {
-		items := table.items.Load()
-		if head > items {
-			head = items
+	repair := func(tables map[string]*freezerTable) (uint64, uint64, error) {
+		var (
+			head = uint64(math.MaxUint64)
+			tail = uint64(0)
+		)
+		for _, table := range tables {
+			head = min(head, table.items.Load())
+			tail = max(tail, table.itemHidden.Load())
 		}
-		hidden := table.itemHidden.Load()
-		if hidden > tail {
-			tail = hidden
+		for _, table := range tables {
+			if err := table.truncateHead(head); err != nil {
+				return 0, 0, err
+			}
+			if err := table.truncateTail(tail); err != nil {
+				return 0, 0, err
+			}
 		}
+		return head, tail, nil
 	}
-	for _, table := range f.tables {
-		if err := table.truncateHead(head); err != nil {
-			return err
-		}
-		if err := table.truncateTail(tail); err != nil {
+
+	headTables, otherTables := splitTables(f.tables)
+	// verify that all the non-header tables have the same size
+	head, tail, err := repair(otherTables)
+	if err != nil {
+		return err
+	}
+	// verify that the header tables have the same size
+	if len(headTables) != 0 {
+		head, tail, err = repair(headTables)
```

Comment:
We still need to make sure the head of all tables are aligned.

This guarantee is certainly broken with this change.

Reply:
Yep done, I also added a change that felix suggested to mark tables as prunable early on